rust/tw_bitcoin/src/recipient_impl.rs

```diff
@@ -0,0 +1,156 @@
+use std::str::FromStr;
+
+use crate::{tweak_pubkey, Error, Recipient, Result};
+use bitcoin::key::{KeyPair, PublicKey, TweakedPublicKey};
+use bitcoin::{PubkeyHash, WPubkeyHash};
+
+impl Recipient<PubkeyHash> {
+    pub fn pubkey_hash(&self) -> &PubkeyHash {
+        &self.t
+    }
+}
+
+impl Recipient<WPubkeyHash> {
+    pub fn from_pubkey_slice(slice: &[u8]) -> Result<Self> {
+        Ok(Recipient {
+            t: PublicKey::from_slice(slice)
+                .map_err(|_| Error::Todo)?
+                .wpubkey_hash()
+                .ok_or(Error::Todo)?,
+        })
+    }
+    pub fn wpubkey_hash(&self) -> &WPubkeyHash {
+        &self.t
+    }
+}
+
+impl Recipient<PublicKey> {
+    pub fn from_slice(slice: &[u8]) -> Result<Self> {
+        Ok(Recipient {
+            t: PublicKey::from_slice(slice).map_err(|_| Error::Todo)?,
+        })
+    }
+}
+
+impl FromStr for Recipient<PublicKey> {
+    type Err = Error;
+
+    fn from_str(string: &str) -> Result<Self> {
+        Self::from_slice(string.as_bytes())
+    }
+}
+
+impl From<PublicKey> for Recipient<PublicKey> {
+    fn from(t: PublicKey) -> Self {
+        Recipient { t }
+    }
+}
+
+impl From<KeyPair> for Recipient<PublicKey> {
+    fn from(keypair: KeyPair) -> Self {
+        Self::from(&keypair)
+    }
+}
+
+impl From<&KeyPair> for Recipient<PublicKey> {
+    fn from(keypair: &KeyPair) -> Self {
+        Recipient {
+            t: PublicKey::new(keypair.public_key()),
+        }
+    }
+}
+
+impl From<PublicKey> for Recipient<TweakedPublicKey> {
```

Comment:
Do we use `Recipient< TweakedPublicKey>`, `Recipient< WPubkeyHash>`, `Recipient< Taproot...>` somewhere?
If not, probably we could remove them?

Reply:
It's kind of useful given that you can easily convert between `Keypair > PublicKey > PubkeyHash/WPubkeyHash`, given that each output type requires different type of data. I can probably simplify this, though.

Reply:
It's kind of useful given that you can easily convert between `Keypair > PublicKey > PubkeyHash/WPubkeyHash`, given that each output type requires different type of data. I can probably simplify this, though.